File: apps/pm/src/data/reminders.ts
/**
 * What falls due next, per property (GET /v1/ops/reminders, #337).
 *
 * The API answers soonest first over the whole book; the grouping is here
 * because a manager works building by building, and a flat list of forty rows
 * from eleven buildings is a list nobody finishes reading.
 */

import { useEffect, useMemo, useState } from 'react';
import { apiFromEnv } from '@dwellm8/mobile-shared';
import type { OpsReminder } from '@dwellm8/mobile-shared';

export type ReminderGroup = {
  id: string;
  name: string;
  locality?: string;
  reminders: OpsReminder[];
};

export type RemindersView = {
  loading: boolean;
  error?: string;
  properties: ReminderGroup[];
  duePaise: number;
  overduePaise: number;
  endingCount: number;
};

const empty: RemindersView = {
  loading: false, properties: [], duePaise: 0, overduePaise: 0, endingCount: 0,
};

export function useReminders(days?: number): RemindersView {
  const api = useMemo(() => apiFromEnv(), []);
  const [rows, setRows] = useState<OpsReminder[] | null>(null);
  const [state, setState] = useState<{ loading: boolean; error?: string }>({ loading: Boolean(api) });

  useEffect(() => {
    if (!api) {
      setState({ loading: false, error: 'The API is not configured on this build.' });
      return;
    }
    let alive = true;
    setState({ loading: true });
    api.opsReminders(days)
      .then((list) => { if (alive) { setRows(list); setState({ loading: false }); } })
      .catch((err: Error) => { if (alive) { setRows(null); setState({ loading: false, error: err.message }); } });
    return () => { alive = false; };
  }, [api, days]);

  return useMemo(() => {
    if (!rows) return { ...empty, loading: state.loading, error: state.error };

    const groups = new Map<string, ReminderGroup>();
    for (const r of rows) {
      const id = r.property_id || r.property;
      const group = groups.get(id)
        ?? { id, name: r.property, locality: r.locality, reminders: [] };
      group.reminders.push(r);
      groups.set(id, group);
    }
    for (const group of groups.values()) {
      group.reminders.sort((a, b) => a.on.localeCompare(b.on));
    }
    const properties = [...groups.values()].sort(
      (a, b) => a.reminders[0].on.localeCompare(b.reminders[0].on));

    return {
      loading: state.loading,
      error: state.error,
      properties,
      duePaise: total(rows, 'rent_due'),
      overduePaise: total(rows, 'rent_overdue'),
      endingCount: rows.filter((r) => r.kind === 'tenancy_ending').length,
    };
  }, [rows, state]);
}

function total(rows: OpsReminder[], kind: OpsReminder['kind']): number {
  return rows.reduce((sum, r) => (r.kind === kind ? sum + (r.amount_minor ?? 0) : sum), 0);
}

File: services/api/internal/surface/ops/reminders.go
package ops

import (
	"net/http"
	"sort"
	"strconv"

	"github.com/tesserix/dwellm8/services/api/internal/platform/effective"
)

// What is about to happen on the book (#337).
//
// The manager's other lists answer for the past: arrears is who already owes,
// today is what the book is worth this morning. Neither says that the rent on
// eleven flats falls due next Tuesday, or that a term runs out inside its own
// notice period — and chasing money after it is late is the thing the reminder
// ladder exists to avoid.

const (
	defaultHorizon = 30
	maxHorizon     = 365
)

// A reminder is one dated thing a manager should act on before it happens.
type reminderResponse struct {
	Kind       string `json:"kind"`
	LeaseID    string `json:"lease_id"`
	PropertyID string `json:"property_id,omitempty"`
	Property   string `json:"property"`
	Unit       string `json:"unit"`
	Locality   string `json:"locality,omitempty"`
	// On is the day it falls due or the day the term ends. Overdue rent is
	// dated today: it is what to do now, and sorting it anywhere else would put
	// the money already lost below the money not yet owed.
	On          string `json:"on"`
	DaysAway    int    `json:"days_away"`
	AmountMinor int64  `json:"amount_minor,omitempty"`
	// InsideNoticeWindow says notice can still be served in time — the fact a
	// renewal decision turns on.
	InsideNoticeWindow bool `json:"inside_notice_window,omitempty"`
}

// Reminders is the manager's forward view: rent falling due, rent already
// overdue, and tenancies running out, over the whole book, soonest first.
func (h *Handler) Reminders(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	today := effective.DateOf(h.now(), ist)
	days := horizon(r.URL.Query().Get("days"))
	through := today.AddDays(days)

	due, err := h.leases.Upcoming(ctx, today, through, rosterLimit)
	if err != nil {
		h.log.Error("reading what falls due", "error", err)
		writeError(w, http.StatusInternalServerError, "could not read the reminders")
		return
	}
	owing, err := h.statements.Outstanding(ctx)
	if err != nil {
		h.log.Error("reading what the firm is owed", "error", err)
		writeError(w, http.StatusInternalServerError, "could not read the reminders")
		return
	}
	ending, err := h.leases.Expiring(ctx, days)
	if err != nil {
		h.log.Error("reading the tenancies running out", "error", err)
		writeError(w, http.StatusInternalServerError, "could not read the reminders")
		return
	}

	out := make([]reminderResponse, 0, len(due)+len(owing)+len(ending))
	for _, d := range due {
		out = append(out, reminderResponse{
			Kind: "rent_due", LeaseID: d.LeaseID, PropertyID: d.PropertyID,
			Property: d.PropertyName, Unit: d.UnitCode, Locality: d.Locality,
			On: d.DueOn.String(), DaysAway: daysBetween(today, d.DueOn), AmountMinor: d.AmountMinor,
		})
	}

	// Where the overdue and the ending tenancies are needs one read each. The
	// debts are already capped by Outstanding's own order, and a tenancy ending
	// is bounded by the horizon, so neither list walks the book.
	if len(owing) > rosterLimit {
		owing = owing[:rosterLimit]
	}
	for _, d := range owing {
		t, err := h.leases.Tenancy(ctx, d.LeaseID, today)
		if err != nil {
			h.log.Warn("describing an overdue tenancy", "lease", d.LeaseID, "error", err)
			continue
		}
		out = append(out, reminderResponse{
			Kind: "rent_overdue", LeaseID: d.LeaseID, PropertyID: t.PropertyID,
			Property: t.PropertyName, Unit: t.UnitCode, Locality: t.Locality,
			On: today.String(), DaysAway: 0, AmountMinor: int64(d.Due),
		})
	}
	for _, e := range ending {
		t, err := h.leases.Tenancy(ctx, e.LeaseID, today)
		if err != nil {
			h.log.Warn("describing a tenancy running out", "lease", e.LeaseID, "error", err)
			continue
		}
		out = append(out, reminderResponse{
			Kind: "tenancy_ending", LeaseID: e.LeaseID, PropertyID: t.PropertyID,
			Property: t.PropertyName, Unit: t.UnitCode, Locality: t.Locality,
			On: e.EndsOn.String(), DaysAway: e.DaysRemaining,
			InsideNoticeWindow: e.InsideNoticeWindow,
		})
	}

	sort.SliceStable(out, func(i, j int) bool { return out[i].On < out[j].On })
	writeJSON(w, http.StatusOK, map[string]any{
		"as_of": today.String(), "horizon_days": days, "reminders": out,
	})
}

// daysBetween counts whole days from one date to another.
func daysBetween(from, to effective.Date) int {
	return int(to.Time().Sub(from.Time()).Hours() / 24)
}

// horizon reads how far ahead to look, bounded: a window of a decade is a
// reminder list nobody reads, and one of zero days is an empty screen.
func horizon(q string) int {
	days, err := strconv.Atoi(q)
	if err != nil || days <= 0 {
		return defaultHorizon
	}
	if days > maxHorizon {
		return maxHorizon
	}
	return days
}
